app/main.py

```diff
@@ -1,2 +1,132 @@
+from dataclasses import dataclass
+from fractions import Fraction
+from typing import Hashable, Any, Iterator, Optional, Iterable
+
+
 class Dictionary:
-    pass
+    @dataclass
+    class Node:
+        key: Hashable
+        obj_hash: hash
+        value: Any
+
+    def __init__(self) -> None:
+        self.capacity = 8
+        self.size = 0
+        self.hash_table: list[Dictionary.Node | None] = [None] * self.capacity
+
+    def _calculate_index(self, key: Hashable) -> int:
+        index = hash(key) % self.capacity
+
+        while (
+            self.hash_table[index]
+            and self.hash_table[index].key != key
+        ):
+            index = (index + 1) % self.capacity
+
+        return index
+
+    @property
+    def current_max_size(self) -> Fraction:
+        return self.capacity * Fraction(2 , 3)
+
+    def _resize(self) -> None:
+        old_hash_table = self.hash_table
+        self.capacity *= 2
+        self.size = 0
+        self.hash_table = [None] * self.capacity
+        for node in old_hash_table:
+            if node:
+                self[node.key] = node.value
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        if self.size > self.current_max_size:
+            self._resize()
+        index = self._calculate_index(key)
+        if not self.hash_table[index]:
+            self.size += 1
+        self.hash_table[index] = Dictionary.Node(key, hash(key), value)
+
+    def __getitem__(self, item: Any) -> Any:
+        index = self._calculate_index(item)
+
+        if self.hash_table[index] is None:
+            raise KeyError(f"Cannot find object with key -> {item}")
+
+        return self.hash_table[index].value
+
+    def __delitem__(self, key: Hashable) -> None:
+        index = self._calculate_index(key)
+
+        if not self.hash_table[index]:
+            raise KeyError(f"Cannot find value for key: {key}")
+
+        self.hash_table[index] = None
+        self.size -= 1
+        if self.size <= self.current_max_size // 2:
```

Comment:
so you don't need to actually change capacity of table when element is deleted. You can make a small change to resize method os it can accept multiplier parameter. And so when it's actual resize you pass there a multiplier of 2 and when it is resize due to element deletion then run it with multiplier of 1